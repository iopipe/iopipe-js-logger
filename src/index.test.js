import path from 'path';
import fs from 'fs-extra';
import _ from 'lodash';
import spawn from 'cross-spawn';

const loggerPlugin = require('.');

beforeEach(() => {
  delete process.env.IOPIPE_LOGGER_ENABLED;
});

test('Can instantiate the plugin with no options', () => {
  const plugin = loggerPlugin();
  const inst = plugin({});
  expect(_.isFunction(inst.hooks['post:setup'])).toBe(true);
  expect(_.isFunction(inst.postSetup)).toBe(true);
  expect(_.isFunction(inst.hooks['post:invoke'])).toBe(true);
  expect(_.isFunction(inst.postInvoke)).toBe(true);
  expect(_.isPlainObject(inst.config)).toBe(true);
  expect(Array.isArray(inst.logs)).toBe(true);
  expect(inst.config.enabled).toBe(false);
});

test('Can enable via object', () => {
  const plugin = loggerPlugin({ enabled: true });
  const inst = plugin();
  expect(inst.config.enabled).toBe(true);
});

test('Can enable via env var', () => {
  process.env.IOPIPE_LOGGER_ENABLED = 'true';
  const plugin = loggerPlugin();
  const inst = plugin();
  expect(inst.config.enabled).toBe(true);
});

test('Can disable via env var', () => {
  process.env.IOPIPE_LOGGER_ENABLED = 'false';
  const plugin = loggerPlugin();
  const inst = plugin();
  expect(inst.config.enabled).toBe(false);
});

test('Works in testProject', () => {
  const resultPath = path.join(__dirname, '../testProject/result.json');
  fs.removeSync(resultPath);
  const result = spawn.sync('node', ['testProject/index.js']);
  const [, stdout, stderr] = result.output;

  // we should still expect that typical console methods report to stdout and stderr while simultaneously capturing them to send in the report
  expect(stdout.toString().split('\n')[0]).toEqual('log-string-test 72');
  expect(stderr.toString()).toMatchSnapshot();

  const { val, logs, putData } = fs.readJsonSync(resultPath);
  expect(val).toEqual('lambda-complete');

  // expect that the intercepted file to S3 has good data in it
  expect(putData[0].startsWith('{"message":"log-string-test 72')).toBeTruthy();

  // ensure the logs have a timestamp now before we strip them out so we don't have a continual conflicting snapshot
  expect(logs[0].timestamp).toBeGreaterThan(1530565119836);
  const logsWithoutTimestamp = logs.map(l => _.omit(l, ['timestamp']));
  expect(logsWithoutTimestamp).toMatchSnapshot();
});
