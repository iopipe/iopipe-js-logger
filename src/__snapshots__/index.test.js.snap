// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Works in testProject 1`] = `
"warn-string-test 72
[Function: hrtime] { showHidden: true }
error-string-test 72
[Function: hrtime] { showHidden: true }
"
`;

exports[`Works in testProject 2`] = `
Array [
  Object {
    "message": "log-string-test 72
",
    "severity": "log",
  },
  Object {
    "message": "[Function: hrtime] { showHidden: true }
",
    "severity": "log",
  },
  Object {
    "message": "debug-string-test 72
",
    "severity": "debug",
  },
  Object {
    "message": "[Function: hrtime] { showHidden: true }
",
    "severity": "debug",
  },
  Object {
    "message": "info-string-test 72
",
    "severity": "info",
  },
  Object {
    "message": "[Function: hrtime] { showHidden: true }
",
    "severity": "info",
  },
  Object {
    "message": "warn-string-test 72
",
    "severity": "warn",
  },
  Object {
    "message": "[Function: hrtime] { showHidden: true }
",
    "severity": "warn",
  },
  Object {
    "message": "error-string-test 72
",
    "severity": "error",
  },
  Object {
    "message": "[Function: hrtime] { showHidden: true }
",
    "severity": "error",
  },
  Object {
    "message": "'dir-string-test'
",
    "severity": "dir",
  },
  Object {
    "message": "{ [Function: hrtime]
  [length]: 1,
  [name]: 'hrtime',
  [prototype]: hrtime { [constructor]: [Circular] } }
",
    "severity": "dir",
  },
]
`;
